rust/yarp/src/lib.rs

```
@@ -0,0 +1,254 @@
+    #[must_use]
+    pub fn source(&self) -> String {
+        self.source.to_str().unwrap().to_string()
+    }
+
+    /// Returns an iterator that can be used to iterate over the errors in the
+    /// parse result.
+    #[must_use]
+    pub fn errors_iter(&self) -> ListIter<Diagnostic> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).error_list;
+            ListIter { current: Diagnostic(list.head.cast::<yp_diagnostic_t>()) }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the warnings in the
+    /// parse result.
+    #[must_use]
+    pub fn warnings_iter(&self) -> ListIter<Diagnostic> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).warning_list;
+            ListIter { current: Diagnostic(list.head.cast::<yp_diagnostic_t>()) }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the comments in the
+    /// parse result.
+    #[must_use]
+    pub fn comments_iter(&self) -> ListIter<Comment> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).comment_list;
+            ListIter { current: Comment(list.head.cast::<yp_comment_t>()) }
+        }
+    }
+
+    /// Returns the root node of the parse result.
+    #[must_use]
+    pub fn node(&self) -> Node {
+        Node::new(self.node.as_ptr())
+    }
+}
+
+impl Drop for ParseResult {
+    fn drop(&mut self) {
+        unsafe {
+            yp_node_destroy(self.parser.as_ptr(), self.node.as_ptr());
+
+            // TODO: Why isn't this working?
+            // yp_parser_free(self.parser.as_ptr());
+        }
+    }
+}
+
+/// Parses the given source string and returns a parse result.
+/// 
+/// # Panics
+/// 
+/// Panics if the source string is not able to be converted into a `CString`.
+/// 
+#[must_use]
+pub fn parse(source: String) -> ParseResult {
+    unsafe {
+        let source = CString::new(source).unwrap();
```

This means passing `String::new("abc\0xyz")` to this function will panic. This should either return an error or use the underlying yarp API that accepts a pointer + a length.